dev-guide/api-conventions.md

```diff
@@ -0,0 +1,257 @@
+
+## API Author Guidance
+
+### Write User Readable Documentation in Godoc
+
+Godoc text is generated into both Swagger and OpenAPI schemas which are then used
+in tools such as `oc explain` or the API reference section of the OpenShift product
+documentation to provide users descriptions of how to use and interact with our products.
+
+In general, Godoc comments should be complete sentences and as much as possible, should
+adhere to the OpenShift product docs [style guide](https://redhat-documentation.github.io/supplementary-style-guide/#style-guidelines).
+
+The Godoc on any field in our API should be sufficiently explained such that an end user
+understands the following:
+
+* What is the purpose of this field? What does it allow them to achieve?
+* How does setting this field interact with other fields or features?
+* What are the limitations of this field?
+  * Does it have any maximum or minimum value?
+  * If it is a string value, are the values limited to a specific list or can it be free form,
+  must it meet a certain regex?
+  * Limitations should be written out within the Godoc as well as added within `kubebuilder` tags. Kubebuilder tags are
+  used for validation but are *not* included within any generated documentation.
+  * See the [validation docs](https://book.kubebuilder.io/reference/markers/crd-validation.html) for inspiration on more validations to apply.
+* Is the field optional or required?
+* When optional, what happens when the field is omitted?
+  * You may choose to set a default value within the API or have a controller default the value
+    at runtime
+
+For example:
+
+```go
+// Example enables developers to understand how to write user facing documentation
+// within the Godoc of their API types.
+// Example is used within the wider Conventions to improve the end user experience
+// and is a required convention.
+// At least one value must be provided within the example and the type should be set
+// appropriately.
+// +kubebuilder:validation:Required
+// + ---
+// + Note that this comment line will not end up in the generated API schema as it is
+// + preceded by a `+`. The `---` also prevents anything after it from being added to
+// + the swagger docs.
+// + This can be used to add notes for developers that aren't intended for end users.
+type Example struct {
+  // Type allows the user to determine how to interpret the example given.
+  // It must be set to one of the following values; Documentation, Convention, or Mixed.
+  // +kubebuilder:validation:Enum:Documentation;Convention;Mixed
+  // +kubebuilder:validation:Required
+  Type string `json:"type"`
+
+  // Documentation allows the user to define documentation for the example.
+  // When this value is provided, the `type` must be set to either `Documentation` or `Mixed`.
+  // The content of the documentation is free form text but must be no longer than 512 characters.
+  // +kubebuilder:validation:MaxLength:=512
```

Comment:
I've noticed that we sometimes use `:=` and sometimes use `=`, and I haven't been able to find any documentation on the distinction.  As far as I can tell, they are interchangeable in practice, and if that is true, I would prefer we settle on a preferred style.  We can leave that for a later PR if it requires debate.